### PR TITLE
Conditionally use rsgeo::line_segmentize() if available

### LINKED ISSUE #522: Speed-up `rnet_merge()`
Two options come to mind for doing this:

- `qgisprocess` for `v.split`, or
- `geos` package (although no direct v.split function)
- `rsgeo` function from @JosiahParry (if ready ; )

Challenge for you if you want it @wangzhao0217: install the `qgisprocess` package + https://trac.osgeo.org/osgeo4w/ and try running the code in the readme of `qgisprocess` pkg.

### PR DESCRIPTION
**Pull request by JosiahParry**
Closes #522. 

This PR speeds up `line_segment()` by utilizing `rsgeo`. This only occurs when the package is installed (there is no prompting introduced in this PR as might be done using `rlang::check_installed()`), and the input geometry either has a missing CRS or is not a geographic CRS this is because rsgeo utilizes euclidean distance to segmentize. If there is a desire to use geodesic / haversine distance in a line_segmentize function, theres potential to add that upstream. However, its my understanding that most road network anlaysis is done on a fairly localized level which should be done using a projected CRS anyways! :) 

Further, this PR converts `line_segment()` into an S3 generic with an `sf` and `sfc_LINESTRING` method enabling people to work with only their geometries and not always have to have a data frame. 

rsgeo is added to the DESCRIPTION's `Suggests` field.

Benchmark: 

```r
library(stplanr)
l <- routes_fast_sf[which(n_vertices(routes_fast_sf) != 2), ] 

bench::mark(
  planar = line_segment(sf::st_set_crs(l, NA), 5),
  og = line_segment(l, 5),
  check = FALSE
)
#> Linking to GEOS 3.11.0, GDAL 3.5.3, PROJ 9.1.0; sf_use_s2() is TRUE
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 planar       2.52ms   2.78ms    271.      1.62MB     37.9
#> 2 og         356.14ms 358.39ms      2.79      30MB     14.0
```

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Amazing. Will check it out now!

File: R/linefuns.R [hunk 1]
```diff
@@ -167,13 +167,70 @@ line_midpoint <- function(l, tolerance = NULL) {
 #' l <- routes_fast_sf[2:4, ]
 #' l_seg_multi = line_segment(l, segment_length = 1000)
 #' plot(sf::st_geometry(l_seg_multi), col = seq(nrow(l_seg_100)), lwd = 5)
-line_segment <- function(
+line_segment <- function(l, n_segments = NA, segment_length = NA) {
+  UseMethod("line_segment")
```
Comment (Robinlovelace) [inline R/linefuns.R:171]: Woo a new generic.

File: R/linefuns.R [hunk 2]
```diff
@@ -210,6 +265,18 @@ line_segment <- function(
 }
 
 
+#' @export
+line_segment.sfc_LINESTRING <- function(
```
Comment (Robinlovelace) [inline R/linefuns.R:269]: Clever.

File: R/linefuns.R [hunk 3]
```diff
@@ -167,13 +167,70 @@ line_midpoint <- function(l, tolerance = NULL) {
...
+
+  if (is.na(n_segments) && is.na(segment_length)) {
+    rlang::abort(
+      "`n_segment` or `segment_length` must be set.",
+      call = rlang::caller_env()
+    )
+  }
+
+  # if rsgeo is available use it
+  if (rlang::is_installed("rsgeo", version = "0.1.6")) {
```
Comment (Robinlovelace) [inline R/linefuns.R:189]: New function for me. Glad to see this is the minimum not pinned version.

File: R/linefuns.R [hunk 4]
```diff
@@ -167,13 +167,70 @@ line_midpoint <- function(l, tolerance = NULL) {
...
+      if (is.na(n_segments)) {
+        l_length <- rsgeo::length_euclidean(geo)
+        n_segments <- max(round(l_length / segment_length), 1)
+      }
+
+      # segmentize the line strings
+      res <- rsgeo::line_segmentize(geo, n_segments)
+
+      # make them into sfc_LINESTRING
+      res <- sf::st_cast(sf::st_as_sfc(res), "LINESTRING")
```
Comment (Robinlovelace) [inline R/linefuns.R:211]: What is it without this casting process, a multilinestring?

**Review by Robinlovelace — APPROVED** [review]
Looking good! Will have a quick play on this branch and maybe add another example or 2 but overall, huge :+1: and big thanks.

File: R/linefuns.R (hunk 4 above)
Comment (JosiahParry) [reply R/linefuns.R:211]: Correct

**Comment by Robinlovelace** [comment]
Straight back at ya, minor changes suggested including exposing `use_rsgeo`, will be useful for benchmarking: https://github.com/JosiahParry/stplanr/pull/1/files

**Comment by Robinlovelace** [comment]
For some reason I get this note, not sure if it's new, but should add to the global allowed names I'm thinking to make checks happier:

```
  Undefined global functions or variables:
    segment
```

**Comment by Robinlovelace** [comment]
That's after my changes...

**Comment by JosiahParry** [comment]
> For some reason I get this note, not sure if it's new, but should add to the global allowed names I'm thinking to make checks happier:
> 
> 
> 
> ```
> 
>   Undefined global functions or variables:
> 
>     segment
> 
> ```

This is likely because there is a variable declared an unused? I won't be able to review for a few more hours.

**Comment by Robinlovelace** [comment]
Merged :+1:

**Pull request merged**